### ISSUE TITLE
[Bug] Auth requires GameKit.framework, which may lead to App store rejection

### ISSUE THREAD

**Issue by sparrow06078** [firebase/quickstart-unity#954]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.1.10f1
  * Firebase Unity SDK version: 7.0.2
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component:Auth
  * Other Firebase Components in use:Auth
  * Additional SDKs you are using:  Facebook SDK Unity SDK 9.0.0
  * Platform you are using the Unity editor on: Windows
  * Platform you are targeting: iOS
  * Scripting Runtime: 

### [REQUIRED] Please describe the issue here:

(Please list the full steps to reproduce the issue. Include device logs, Unity logs, and stack traces if available.)

#### Steps to reproduce:

Have you been able to reproduce this issue with just the Firebase Unity quickstarts (this GitHub project)?
What's the issue repro rate? (eg 100%, 1/5 etc)

What happened? How can we make the problem occur?
This could be a description, log/console output, etc.

If you have a downloadable sample project that reproduces the bug you're reporting, you will
likely receive a faster response on your issue.

#### Relevant Code:

Engine : Unity 2020.1.10f1
Using Firebase and Facebook SDK Unity SDK 9.0.0
Build in iOS and Xcode 12.4

I have two error during build in Xcode after add Facebook  SDK

1.

'FBSDKCoreKit/FBSDKCoreKit.h' file not found

This error just happened at .xcodeproj
I don’t know why but It didn’t happened if I builded it at .xcworkspace
But even at .xcworkspace , I still faced the error 2 

2.

Undefined symbol: _OBJC_CLASS_$_GKLocalPlayer

I already tried this solution :

adding GameKit.framework to your xcode prject from
Targets -> UnityFramework -> General -> Frameworks and Libraries
-> click + button and find GameKit.framework -> rebuild.

But it didn’t work for me.
Also try to add 
libFirebaseCppAuth.a in Compile Sources . But it still failed to build .
Please help me if you know how to solve it.

**Comment by cynthiajoan** [comment]
Hi @sparrow06078,

You mentioned you've tried solution I guess reference from this [comment](https://github.com/firebase/quickstart-unity/issues/862#issuecomment-769591776), which I also see in a following [one](https://github.com/firebase/quickstart-unity/issues/862#issuecomment-769741637) it mentioned that works with a clean first and then build. Could you try again to see if that can also help you?

**Comment by sparrow06078** [comment]
> Hi @sparrow06078,
> 
> You mentioned you've tried solution I guess reference from this [comment](https://github.com/firebase/quickstart-unity/issues/862#issuecomment-769591776), which I also see in a following [one](https://github.com/firebase/quickstart-unity/issues/862#issuecomment-769741637) it mentioned that works with a clean first and then build. Could you try again to see if that can also help you?

I tried this but didnt work.

Targets -> UnityFramework -> General -> Frameworks and Libraries
-> click + button and find GameKit.framework -> rebuild.

**Comment by alexames** [comment]
`'FBSDKCoreKit/FBSDKCoreKit.h' file not found` sounds like an error with the Facebook SDK you're using. Have you confirmed that using the Facebook SDK works when Firebase is not included? Or the opposite: that Firebase works when Facebook is not included?

`GameKit.framework` also shouldn't affect Firebase SDKs as far as I know. 

Can you try starting from the Unity Quickstart app and see if you can get that working, then add in libraries until it breaks to see what specifically is breaking your build?

**Comment by shlighter** [comment]
I'm having the same issue with XCODE and SDK 7.1.0

Undefined symbol: _OBJC_CLASS_$_GKLocalPlayer  referenced from objc-class-ref in libFirebaseCppAuth.a
ld: symbol not find for architecture arm64

Please fix it!!!

**Comment by chkuang-g** [comment]
@sparrow06078 

I believe your issue with `FBSDKCoreKit/FBSDKCoreKit.h` is a problem in Facebook SDK. See this https://github.com/facebook/facebook-sdk-for-unity/issues/504

Facebook SDK also introduced a number of issue when used with Firebase SDK.  And those should be fixed in the latest `7.1.0`. This includes [a patch in Firebase](https://firebase.google.com/support/release-notes/unity#version_710_-_february_16_2021), and [a couple of new features in EDM4U](https://github.com/googlesamples/unity-jar-resolver/releases/tag/v1.2.164) in order to properly support Swift Framework used by Facebook SDK.

In short, try to upgrade to Firebase `7.1.0` and see if that resolves your issue.

Let us know.
Shawn

**Comment by chkuang-g** [comment]
@shlighter Thanks so much for your information provided in #981

Here is my understanding, combining information from #981 and https://github.com/facebook/facebook-sdk-for-unity/issues/504#issuecomment-746729238

1. When using Firebase SDK and Facebook SDK, you will get `XCODE ERROR- OBJC_CLASS$_GKLocalPlayer in libFirebaseCppAuth.a symbol not find for architecture arm64` when building Xcode project.
   - I actually tried quickstart and import both Firebase `7.1.0` and Facebook `9.0.0` and I did not see this build error at all. 
   - @shlighter reported that they only use Firebase SDK and see the same error.  Therefore, this is not necessary to be related to Facebook SDK.
   - The common workaround is to add `GameKit.framework` to `UnityFramework` target.  However, this would cause Apple App Store reject if the app did not GameCenter feature at all. In that case, the only workaround is to modify iOS code using instruction described here https://github.com/facebook/facebook-sdk-for-unity/issues/504#issuecomment-746729238
1. I did noticed that Auth C++ code has dependency to GameKit [here](https://github.com/firebase/firebase-cpp-sdk/blob/e86aa624a97ed172056fc76a4902ab6257f3f8bc/auth/src/ios/credential_ios.mm#L172).  I suspect the link error is from here.  I think we probably should use [Runtime Reference](https://developer.apple.com/documentation/objectivec/objective-c_runtime), a.k.a. Reflection for Objective-C, here.

Seems to be a legit issue.  Let me report to the team.

**Comment by shlighter** [comment]
Hi there, I see you closed my ticket and continue here. I want to add that in my case I don't have Facebook SDK at all, I only have Firebase

**Comment by dylangolow** [comment]
I also have this issue and have had to look into PostProcessingScripts to resolve this. Our team is using the Firebase SDK. Just adding this to the list so we can hopefully have a fix for this instead of requiring "unused" frameworks to be added to the build.

**Comment by kjyv** [comment]
We also have this, no Facebook SDK. This happened after updating from 7.0.2 to 7.1.0. We have to add GameKit.framework to the Xcode project.

**Comment by realworld666** [comment]
I'm having the same issue too. We have also just updated to the 7.1.0 SDK but we don't use Facebook and never have.

**Comment by mastef** [comment]
Also ran into this issue with `GKLocalPlayer`. Have both Firebase and Facebook.

**Comment by shlighter** [comment]
2 months and no solution still? It used to work fine before the latest SDK releases

**Comment by SleepyWolfy** [comment]
Bump

**Comment by safaiyeh** [comment]
This comment here helped me https://github.com/facebook/facebook-sdk-for-unity/issues/504#issuecomment-747153904

**Comment by safaiyeh** [comment]
The Unity framework is simply missing the dependency to GameKit that Firebase requires. There should be some separation between the Game Center auth modules and the rest.

**Comment by george-s-mindo** [comment]
Having same issue when started to use FirebaseAuth on my unity project. Thankfully, use of this comment helped to build iOS project: https://github.com/firebase/quickstart-unity/issues/954#issuecomment-780320439. This is major issue, as it takes around 2 hours to find the rootcause. Bump bump bump

**Comment by TomaskoKnaze** [comment]
I'm also having the same issue with GKLocalPlayer on 7.1.0 SDK, not using facebook.

**Comment by uladzislaubasin** [comment]
This issue should not be closed, because the problem is fixed in a separate package and the latest Unity package still does not work. Please, reopen it

**Comment by streetwalker-unity** [comment]
I'm not using Facebook - just the Apple Sign-in Authorization ( i've in included Firebase Auth in my Unity project) and the XCode build crashes with:  [...] FirebaseUnity.app/Frameworks/UnityFramework.framework/UnityFramework, 265): Symbol not found: _OBJC_CLASS_$_GKLocalPlayer

So apparently GameKit is still required.   How do I go about resolving this so I can get the build working in XCode?

**Comment by uladzislaubasin** [comment]
@streetwalker-unity this is because the issue not in Facebook SDK, but in Firebase Auth.
This workaround works for me: https://github.com/facebook/facebook-sdk-for-unity/issues/504#issuecomment-747016544
No issues with AppStore review. Game has been successfully published there: https://apps.apple.com/us/app/bitten/id1556407553

Just a kind reminder for this project owners to fix this issue and update Firebase SDK.

**Comment by streetwalker-unity** [comment]
@uladzislaubasin  - so what part gets you past, or you have to perform all of these?
1.  Enable GameKit.framework as optional in the UnityFramework target, Link Libraries in the Build Settings
2. Edit the FIRGameCenterAuthProvider.m an insert the code provided
3. Add the C# script provided to the Editors folder (somewhere...)

Did you have to perform all 3 of those?

**Comment by uladzislaubasin** [comment]
Only 1 and 2.

Point 3 is to automate actions from 1, otherwise you have to repeat 1 and 2 every time you export FRESH iOS build. I did not take this path, because I hope that Firebase SDK will be updated soon and this workaround won't be required anymore.

**Comment by artoonie** [comment]
> 1. Enable GameKit.framework as optional in the UnityFramework target, Link Libraries in the Build Settings

This step worked for me, but it's a frustrating workaround as I have to do this every time the xcode project is rebuilt.

I'd also note that it only began happening when I started using https://github.com/googlesamples/google-signin-unity along with firebase auth. Unsure if it's related, or if nothing was being linked previously.

**Comment by safaiyeh** [comment]
You don’t have to always redo it when building the iOS app, you can write a post build script to link it.

How to do it:
https://github.com/facebook/facebook-sdk-for-unity/issues/504#issuecomment-757912373

**Comment by artoonie** [comment]
Ah, I missed that, thank you!
